**Assignment 9**  
Matt Caraher

In [38]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import sklearn.metrics as metrics
from sklearn import linear_model
from sklearn import tree
from sklearn import neural_network
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import GridSearchCV

In [2]:
pip install ucimlrepo

Note: you may need to restart the kernel to use updated packages.


1. Performing exploratory analysis on the data and removing motor_UPDRS column.

In [3]:
# Using import code sample from the dataset website
from ucimlrepo import fetch_ucirepo 
  
# fetch dataset 
parkinsons_telemonitoring = fetch_ucirepo(id=189) 
  
# data (as pandas dataframes) 
X = parkinsons_telemonitoring.data.features 
y = parkinsons_telemonitoring.data.targets 
  
# metadata 
print(parkinsons_telemonitoring.metadata) 
  
# variable information 
print(parkinsons_telemonitoring.variables) 


{'uci_id': 189, 'name': 'Parkinsons Telemonitoring', 'repository_url': 'https://archive.ics.uci.edu/dataset/189/parkinsons+telemonitoring', 'data_url': 'https://archive.ics.uci.edu/static/public/189/data.csv', 'abstract': "Oxford Parkinson's Disease Telemonitoring Dataset", 'area': 'Health and Medicine', 'tasks': ['Regression'], 'characteristics': ['Tabular'], 'num_instances': 5875, 'num_features': 19, 'feature_types': ['Integer', 'Real'], 'demographics': ['Age', 'Sex'], 'target_col': ['motor_UPDRS', 'total_UPDRS'], 'index_col': ['subject#'], 'has_missing_values': 'no', 'missing_values_symbol': None, 'year_of_dataset_creation': 2009, 'last_updated': 'Fri Nov 03 2023', 'dataset_doi': '10.24432/C5ZS3N', 'creators': ['Athanasios Tsanas', 'Max Little'], 'intro_paper': {'title': "Accurate Telemonitoring of Parkinson's Disease Progression by Noninvasive Speech Tests", 'authors': 'A. Tsanas, Max A. Little, P. McSharry, L. Ramig', 'published_in': 'IEEE Transactions on Biomedical Engineering', 

In [4]:
print('Exploring the dataset\n')
print('First rows of the dataset')
print(X.head())
print('\nData types')
print(X.dtypes)
print('\nUsing describe on the data')
print(X.describe())
print('\nNull values:', X.isnull().sum().sum())

Exploring the dataset

First rows of the dataset
   age  test_time  Jitter(%)  Jitter(Abs)  Jitter:RAP  Jitter:PPQ5  \
0   72     5.6431    0.00662     0.000034     0.00401      0.00317   
1   72    12.6660    0.00300     0.000017     0.00132      0.00150   
2   72    19.6810    0.00481     0.000025     0.00205      0.00208   
3   72    25.6470    0.00528     0.000027     0.00191      0.00264   
4   72    33.6420    0.00335     0.000020     0.00093      0.00130   

   Jitter:DDP  Shimmer  Shimmer(dB)  Shimmer:APQ3  Shimmer:APQ5  \
0     0.01204  0.02565        0.230       0.01438       0.01309   
1     0.00395  0.02024        0.179       0.00994       0.01072   
2     0.00616  0.01675        0.181       0.00734       0.00844   
3     0.00573  0.02309        0.327       0.01106       0.01265   
4     0.00278  0.01703        0.176       0.00679       0.00929   

   Shimmer:APQ11  Shimmer:DDA       NHR     HNR     RPDE      DFA      PPE  \
0        0.01662      0.04314  0.014290  21.640  

In [5]:
print('Exploring the dataset\n')
print('First rows of the dataset')
print(y.head())
print('\nData types')
print(y.dtypes)
print('\nUsing describe on the data')
print(y.describe())
print('\nNull values:', y.isnull().sum().sum())

Exploring the dataset

First rows of the dataset
   motor_UPDRS  total_UPDRS
0       28.199       34.398
1       28.447       34.894
2       28.695       35.389
3       28.905       35.810
4       29.187       36.375

Data types
motor_UPDRS    float64
total_UPDRS    float64
dtype: object

Using describe on the data
       motor_UPDRS  total_UPDRS
count  5875.000000  5875.000000
mean     21.296229    29.018942
std       8.129282    10.700283
min       5.037700     7.000000
25%      15.000000    21.371000
50%      20.871000    27.576000
75%      27.596500    36.399000
max      39.511000    54.992000

Null values: 0


In [6]:
y = y.drop(columns=['motor_UPDRS'])

In [7]:
print('Exploring the dataset\n')
print('First rows of the dataset')
print(y.head())
print('\nData types')
print(y.dtypes)
print('\nUsing describe on the data')
print(y.describe())
print('\nNull values:', y.isnull().sum().sum())

Exploring the dataset

First rows of the dataset
   total_UPDRS
0       34.398
1       34.894
2       35.389
3       35.810
4       36.375

Data types
total_UPDRS    float64
dtype: object

Using describe on the data
       total_UPDRS
count  5875.000000
mean     29.018942
std      10.700283
min       7.000000
25%      21.371000
50%      27.576000
75%      36.399000
max      54.992000

Null values: 0


2. Using cross-validation to build a linear regression model to predict total_UPDRS

In [28]:
lreg = linear_model.LinearRegression()

lreg_accuracies = cross_val_score(lreg, X, y, cv = 5, 
                             scoring = 'neg_mean_absolute_error')

print(lreg_accuracies)
lreg_accuracies = -1 * lreg_accuracies
print(lreg_accuracies)

[ -9.37029693  -8.24834453 -10.82542378  -8.14486908 -13.29115806]
[ 9.37029693  8.24834453 10.82542378  8.14486908 13.29115806]


3. Using cross-validation to build a regression tree model to predict total_UPDRS

In [29]:
tree_model = tree.DecisionTreeRegressor(min_samples_leaf = 5,
                                       random_state = 16)

tm_accuracies = cross_val_score(tree_model, X, y, cv = 5, 
                             scoring = 'neg_mean_absolute_error')

tm_accuracies = -1 * tm_accuracies
print(tm_accuracies)

[15.26727581 10.92936692  8.73228856 10.24742347 13.19420816]


4. Using cross-validation to build a neural network model to predict total_UPDRS

In [26]:
print('Data preparation for neural network\n')
print(y)
y = y.ravel()
print(y)

Data preparation for neural network

[34.398 34.894 35.389 ... 32.495 32.007 31.513]
[34.398 34.894 35.389 ... 32.495 32.007 31.513]


In [30]:
nn_model = neural_network.MLPRegressor(10, activation = 'logistic',
                                      max_iter = 10000, random_state = 16)

nn_accuracies = cross_val_score(nn_model, X, y, cv = 5,
                            scoring = 'neg_mean_absolute_error')

nn_accuracies = -1 * nn_accuracies
print(nn_accuracies)

[10.65923093  8.66379685 10.01492924  7.58459846 11.42773202]


5. Compare the performance of the 3 models based on MAE

In [37]:
lreg_avg_mae = np.mean(lreg_accuracies)
tm_avg_mae = np.mean(tm_accuracies)
nn_avg_mae = np.mean(nn_accuracies)

print('Linear regression average MAE:', lreg_avg_mae)
print('Tree model average MAE:', tm_avg_mae)
print('Neural network average MAE:', nn_avg_mae)

Linear regression average MAE: 9.976018475061968
Tree model average MAE: 11.674112584356635
Neural network average MAE: 9.670057499506564


Based on the above results, the neural network has the best mean absolute error score, with approximately 9.67. Closely behind is the linear regression model with 9.98, and then the tree model with 11.67. I can optimize this model by use of grid search. 

6. Trying to optimize the neural network model

In [43]:
tuned_params = {'hidden_layer_sizes': [5, 10, 15, 20, 25, 30],
               'activation': ['logistic', 'relu', 'tanh']}

nn = neural_network.MLPRegressor(max_iter = 10000, random_state = 16)

nn_optimizer = GridSearchCV(nn, tuned_params,
                           scoring = 'neg_mean_absolute_error',
                           cv = 5,
                           return_train_score = False,
                           verbose = 0)

nn_optimizer.fit(X, y)

print('Best params:\n', nn_optimizer.best_params_)
print()
print('Best estimator:\n', nn_optimizer.best_estimator_)

Best params:
 {'activation': 'relu', 'hidden_layer_sizes': 25}

Best estimator:
 MLPRegressor(hidden_layer_sizes=25, max_iter=10000, random_state=16)


In [45]:
-nn_optimizer.best_score_

9.180054974184866

with an mae of 9.18, the model has improved from the original, thanks to the use of grid search. 